### PR TITLE
Feature: Cached client requests

### PR DESCRIPTION
**Pull request by tdietert**
This branch was originally created to debug a potential issue in the system where the client was timing out on a write request even though the network should be live (a majority of nodes could exchange messages); however, it has unearthed yet another problem with the current implementation of client interactions.

We originally sought to provide a minimal, feature _in_complete example with `app/Main.hs` and `src/Examples/Raft`; however, we now use that executable for black box testing and along the way we have unearthed some disparities in the way in which clients expect to interact with the system and how the interaction actually occurs.

As mentioned in the ARC Raft report: 
> The Raft authors argue that linearisable semantics for client commands to the state
machine are achieved by assigning all commands a sequential serial number when
committed to the cluster by a client. The state machine then caches the last serial
number for each client along with its corresponding response. This is referred to as
the client request cache. Thus when a client retransmits a request, if the serial number
of the committed command is found in the cache, the command is not reapplied to
the machine and the cached response is returned. This assumes that commands are
committed to the state machine in order of serial number; thus each client is limited
to at most one outstanding request

In the the model testing we are doing (found in `test/QuickCheckStateMachine.hs`), the model would retry up to 20 (arbitrarily chosen) times to submit a read, write, or set command to the current leader during period of node isolation. Sometimes, the model would expect the right number of commands to be executed on the machine, but other times the system would actually acknowledge a subset of the total requests issued during retries. This is because there was no **serial number** attached to the client requests. Sometimes the isolated node to which the client write request was issued thought itself to be leader, and once the node connections were fixed, the node became leader for the new term; **in this case, some of the write requests that timed-out _were actually included in the system log_**. 

Thus, we must implement a client request cache and client request serial numbers because in the event of a client request timeout, the client may issue the request again. In the case that the request timed out, but the system **does** replicate the request eventually, future requests that are retried with the same serial number will be acknowledged with a cached result and _definitely not be replicated multiple times_.

**TLDR**: So that clients don't end up replicated multiple copies of the same command when they meant to only issue a single successful request, we must implement client serial numbers on requests and client request caches on leaders.

This PR _also_ quite dramatically refactors the existing client interface such that all bookkeeping the client needs to do when interacting with the raft system is abstracted, aside from timeouts. See the `src/Raft/Client.hs` module.

### CODE REVIEW

**Comment by tdietert** [comment]
Development will start by referencing section 4.3 of the ARC report:
> Inspired by the approach taken by Viewstamped Replication [26], we moved the
client request cache to the consensus module and abstracted it from the state machine.
For each incoming request to the leader, the leader looks up the client ID in the client
request cache:

> • If the requested serial number is less then the last serial number, we can ignore
the command, as the client must have already received a response.

> • If the requested serial number is the same as the last serial number, we reply
with the cached response if present.

> • If the requested serial number is successor of the last serial number, the command
can be added to the log and the last serial number updated. Later, once
the command is committed to the state machine, the response can be cached
alongside the serial number.

> But how can it be safe to only store the last serial number and response, when entries
can removed from the log? If any node has ever seen a serial number greater than any
others last serial number, then the cached response will never be needed as the client
must have received the response from that committed entry. This positioning of the
client request cache introduced a livelock bug, which we later address (§4.3).
